File: src/routes/channels.ts
import {Elysia} from "elysia";
import isAuthenticated from "../middleware/isAuthenticated";
import {db} from "../db";
import {messages} from "../db/schema/message";
import z from "zod/v4";
import {Snowflake} from "@theinternetfolks/snowflake";
import {and, eq, gt, InferSelectModel, lt} from "drizzle-orm";
import {invites} from "../db/schema/guild";
import * as randomstring from "randomstring";
import rateLimit from "../middleware/rateLimit";
import {ip} from "elysia-ip";

const messageCreateBody = z.object({
    content: z.string().min(1).max(2000),
    nonce: z.string().optional()
});

const messageUpdateBody = z.optional(messageCreateBody);

const messageRetrieveQuery = z.object({
    before: z.string().optional(),
    after: z.string().optional(),
    limit: z.coerce.number().min(1).max(100).default(50),
    around: z.string().optional()
});

const inviteCreateBody = z.object({
    max_uses: z.number().max(100).optional(),
    max_age: z.number().max(604800).default(86400).optional()
});

type GetMessagesOptions = {
    before?: string;
    after?: string;
    around?: string;
    limit: number;
}

async function getMessages(channel: string, options: GetMessagesOptions): Promise<InferSelectModel<typeof messages>[]> {
    if (options.around) {
        const before = await getMessages(channel, {
            before: options.around,
            limit: (options.limit ?? 50) / 2
        });
        const after = await getMessages(channel, {
            after: options.around,
            limit: (options.limit ?? 50) / 2
        });
        return [...before, ...after];
    }

    let q = eq(messages.channelId, channel);
    if (options.before) q = and(q, lt(messages.id, options.before))!;
    else if (options.after) q = and(q, gt(messages.id, options.after))!;


    return (await db.query.messages.findMany({
        limit: (options.limit ?? 50),
        where: q,
        with: {
            author: true
        },
        orderBy: (messages, { desc }) => [desc(messages.createdAt)]
    })).reverse();
}

export default new Elysia({prefix: '/channels'})
    .use(ip())
    .guard({
            beforeHandle(ctx) {
                if (!isAuthenticated(ctx.cookie[Bun.env.NODE_ENV === 'production' ? '__Host-Token' : 'token'])) return ctx.status('Unauthorized');
            }
        }, (_app) =>
            _app
                .group('/:id', (app) =>
                    app
                        .resolve(async (ctx) => {
                            const user = await db.query.accounts.findFirst({
                                where: (accounts, {eq}) => eq(accounts.token, ctx.cookie[Bun.env.NODE_ENV === 'production' ? '__Host-Token' : 'token']!.value!),
                                with: {
                                    user: true
                                }
                            });
                            
                            const channel = await db.query.channels.findFirst({
                                where: (channels, {eq}) => eq(channels.id, ctx.params.id),
                                with: {
                                    guild: true
                                }
                            });
                            
                            if (!channel) {
                                return {user, member: null, channel: null, guild: null};
                            }
                            
                            const guildMember = await db.query.guildMembers.findFirst({
                                where: (guildMembers, {eq, and}) => and(
                                    eq(guildMembers.userId, user!.id),
                                    eq(guildMembers.guildId, channel.guildId)
                                )
                            });

                            if (guildMember) {
                                return { user, member: guildMember, guild: channel.guild, channel };
                            }

                            return {user, member: null, channel, guild: null};
                        })
                        .guard({
                            async beforeHandle(ctx) {
                                if (!ctx.member) {
                                    console.log("no member");
                                    return ctx.status('Unauthorized', {
                                        code: 'messages.errors.unauthorized'
                                    });
                                }
                                if (!ctx.channel) {
                                    console.log("no channel");
                                    // @ts-expect-error
                                    return ctx.status('Not Found', {
                                        code: 'messages.errors.channelNotFound'
                                    });
                                }
                                const { limited, retryAfter } = await rateLimit(
                                    ctx.ip,
                                    ctx.channel!.rateLimitPerUser > 0 ? 1 : 50,
                                    ctx.channel!.rateLimitPerUser > 0 ? ctx.channel!.rateLimitPerUser : 1,
                                    `messages::${ctx.channel!.id}`
                                );
                                if (limited) return ctx.status('Too Many Requests', {
                                    message: 'You are being rate limited.',
                                    retryAfter
                                });
                            }
                        })
                        .post('/messages', async (ctx) => {
                            const data = messageCreateBody.safeParse(JSON.parse(ctx.body as any));
                            if (!data.success) {
                                console.log(data.error);
                                return ctx.status('Bad Request', {
                                    code: 'messages.errors.validationFailed'
                                });
                            }
                            try {
                                const id = Snowflake.generate();
                                const message = await db.insert(messages).values({
                                    content: data.data.content,
                                    authorId: ctx.user!.id,
                                    guildId: ctx.member!.guildId,
                                    channelId: ctx.params.id,
                                    type: 'DEFAULT',
                                    id,
                                    nonce: data.data.nonce ?? "0"
                                }).returning();
                                ctx.server?.publish(ctx.member!.guildId, JSON.stringify({
                                    op: 0,
                                    t: "MESSAGE_CREATE",
                                    d: {
                                        ...message[0],
                                        author: {
                                            ...ctx.user!.user,
                                            member: {
                                                nickname: ctx.member!.nickname
                                            }
                                        }
                                    }
                                }));
                                return {
                                    ...message[0],
                                    author: {
                                        ...ctx.member,
                                        ...ctx.user!.user
                                    }
                                };
                            } catch (e) {
                                console.error(e);
                                return ctx.status('Internal Server Error', {
                                    e
                                });
                            }
                        })
                        .get('/messages', async (ctx) => {
                            const query = messageRetrieveQuery.safeParse(ctx.query);
                            if (!query.success) {
                                console.error(query.error);
                                return ctx.status('Bad Request', {
                                    code: 'errors.validationFailed'
                                });
                            }
                            return await getMessages(ctx.params.id, query.data);
                        })
                        .patch('/messages/:messageId', async (ctx) => {
                            const data = messageUpdateBody.safeParse(JSON.parse(ctx.body as string));
                            if (!data.success) {
                                console.log(data.error);
                                return ctx.status('Bad Request', {
                                    code: 'messages.errors.validationFailed'
                                });
                            }
                            const message = await db.query.messages.findFirst({
                                where: (messages, {eq}) => eq(messages.id, ctx.params.messageId)
                            });
                            if (!message) return ctx.status('Not Found', {
                                code: 'messages.errors.messageNotFound'
                            });
                            if (message.authorId !== ctx.user!.userId) return ctx.status('Forbidden', {
                                code: 'messages.errors.notYourMessage'
                            });
                            if (message.content === data.data?.content) return message;
                            const updatedMessage = await db
                                .update(messages)
                                .set({
                                    content: data.data?.content,
                                    updatedAt: new Date()
                                })
                                .where(eq(messages.id, ctx.params.messageId))
                                .returning();
                            ctx.server?.publish(message.guildId, JSON.stringify({
                                op: 0,
                                t: 'MESSAGE_UPDATE',
                                d: updatedMessage[0]
                            }));
                            return updatedMessage[0];
                        })
                        .delete('/messages/:messageId', async (ctx) => {
                            const message = await db.query.messages.findFirst({
                                where: (messages, {eq}) => eq(messages.id, ctx.params.messageId)
                            });
                            if (!message) return ctx.status('Not Found', {
                                code: 'messages.errors.messageNotFound'
                            });
                            if (ctx.user!.userId !== message.authorId) return ctx.status('Forbidden', {
                                code: 'messages.errors.forbidden'
                            });
                            await db
                                .delete(messages)
                                .where(eq(messages.id, ctx.params.messageId));
                            ctx.server?.publish(message.guildId, JSON.stringify({
                                op: 0,
                                t: "MESSAGE_DELETE",
                                d: {
                                    id: message.id,
                                    channelId: message.channelId,
                                    guildId: message.guildId
                                }
                            }));
                            return ctx.status('No Content');
                        })
                        .post('/typing', async (ctx) => {
                            try {
                                ctx.server?.publish(ctx.member!.guildId, JSON.stringify({
                                    op: 0,
                                    t: "TYPING_START",
                                    d: {
                                        channelId: ctx.params.id,
                                        userId: ctx.user!.userId,
                                        user: {
                                            ...ctx.user!.user,
                                            member: {
                                                nickname: ctx.member!.nickname
                                            }
                                        },
                                        timestamp: Date.now(),
                                        expiresAt: Date.now() + 10000
                                    }
                                }));

                                return { success: true };
                            } catch (e) {
                                console.error(e);
                                return ctx.status('Internal Server Error');
                            }
                        })
                        .post('/invites', async (ctx) => {
                            const query = inviteCreateBody.safeParse(ctx.body);
                            if (!query.success) {
                                console.error(query.error);
                                return ctx.status('Bad Request', {
                                    code: 'errors.validationFailed'
                                });
                            }
                            try {
                                const code = randomstring.generate({
                                    charset: "alphanumeric",
                                    length: 6
                                });
                                const id = Snowflake.generate();
                                return (await db.insert(invites).values({
                                    code,
                                    id,
                                    guildId: ctx.guild!.id,
                                    channelId: ctx.params.id,
                                    creatorId: ctx.user!.id
                                }).returning())[0];
                            } catch (e) {
                                console.error(e);
                                return ctx.status('Internal Server Error');
                            }
                        })
                )
    )

File: src/routes/channels.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::IntoResponse,
    routing::{delete, get, patch, post},
    Json, Router,
};
use axum_extra::extract::CookieJar;
use serde::Deserialize;
use serde_json::json;

use byteorm_client::MessageType;

use crate::auth::token::verify_token;
use crate::error::AppError;
use crate::middleware::rate_limit;
use crate::models::{Invite, Message, User};
use crate::state::SharedState;

#[derive(Debug, Deserialize)]
pub struct MessageCreateRequest {
    content: String,
    nonce: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct MessageUpdateRequest {
    content: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct MessagesQuery {
    before: Option<String>,
    after: Option<String>,
    limit: Option<i64>,
    around: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct InviteCreateRequest {
    max_uses: Option<i32>,
    max_age: Option<i32>,
}

fn generate_snowflake() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as u64;
    
    let epoch = 1420070400000u64;
    let timestamp = now - epoch;
    
    use rand::Rng;
    let random: u64 = rand::thread_rng().gen_range(0..4096);
    
    let id = (timestamp << 22) | random;
    id.to_string()
}

fn generate_code() -> String {
    use rand::Rng;
    let chars: Vec<char> = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        .chars()
        .collect();
    let mut rng = rand::thread_rng();
    (0..6).map(|_| chars[rng.gen_range(0..chars.len())]).collect()
}

pub fn router() -> Router<SharedState> {
    Router::new()
        .route("/{id}/messages", get(get_messages))
        .route("/{id}/messages", post(create_message))
        .route("/{id}/messages/{message_id}", patch(update_message))
        .route("/{id}/messages/{message_id}", delete(delete_message))
        .route("/{id}/typing", post(typing_start))
        .route("/{id}/invites", post(create_invite))
}

async fn get_current_user(
    state: &SharedState,
    jar: &CookieJar,
) -> Result<byteorm_client::Accounts, AppError> {
    let is_production = std::env::var("NODE_ENV").unwrap_or_default() == "production";
    let cookie_name = if is_production { "__Host-Token" } else { "token" };
    
    let token = jar.get(cookie_name)
        .map(|c| c.value().to_string())
        .ok_or(AppError::Unauthorized)?;
    
    let _token_data = verify_token(&token, &state.config.token_secret)
        .ok_or(AppError::Unauthorized)?;
    
    state.db.accounts
        .find_first(|q| q.where_token(token))
        .await?
        .ok_or(AppError::Unauthorized)
}

async fn get_channel_and_member(
    state: &SharedState,
    channel_id: &str,
    user_id: &str,
) -> Result<(byteorm_client::Channels, byteorm_client::GuildMembers), AppError> {
    let channel = state.db.channels
        .find_first(|q| q.where_id(channel_id.to_string()))
        .await?
        .ok_or(AppError::NotFound("messages.errors.channelNotFound".to_string()))?;

    let member = state.db.guild_members
        .find_first(|q| q
            .where_user_id(user_id.to_string())
            .where_guild_id(channel.guild_id.clone()))
        .await?
        .ok_or(AppError::Unauthorized)?;

    Ok((channel, member))
}

async fn get_messages(
    State(state): State<SharedState>,
    jar: CookieJar,
    Path(channel_id): Path<String>,
    Query(query): Query<MessagesQuery>,
) -> Result<impl IntoResponse, AppError> {
    let account = get_current_user(&state, &jar).await?;
    let (_channel, _member) = get_channel_and_member(&state, &channel_id, &account.id).await?;
    
    let limit = query.limit.unwrap_or(50).min(100).max(1) as usize;
    
    let mut builder = state.db.messages.query()
        .where_channel_id(channel_id)
        .include_users()
        .order_by_created_at_desc()
        .limit(limit);

    if let Some(before) = &query.before {
        builder = builder.where_id_lt(before.clone());
    } else if let Some(after) = &query.after {
        builder = builder.where_id_gt(after.clone());
    }
    
    let messages_json = builder.find_many_json().await
        .map_err(|e| AppError::InternalServerError(e.to_string()))?;
    
    let mut messages: Vec<Message> = messages_json.into_iter().map(|value| {
        let msg_model: byteorm_client::Messages = serde_json::from_value(value.clone())
             .expect("Failed to deserialize message");
        
        let user_json = value.get("users").expect("Missing users join");
        let user_model: byteorm_client::Users = serde_json::from_value(user_json.clone())
             .expect("Failed to deserialize user");
             
        let mut msg: Message = msg_model.into();
        let user: User = user_model.into();

        msg.author = Some(user);

        msg
    }).collect();
    
    messages.reverse();
    
    Ok(Json(messages))
}

async fn create_message(
    State(state): State<SharedState>,
    jar: CookieJar,
    Path(channel_id): Path<String>,
    body: axum::body::Bytes,
) -> Result<impl IntoResponse, AppError> {
    let body: MessageCreateRequest = serde_json::from_slice(&body)
        .map_err(|_| AppError::BadRequest("Invalid JSON".to_string()))?;
    let account = get_current_user(&state, &jar).await?;
    let (channel, member) = get_channel_and_member(&state, &channel_id, &account.id).await?;

    let user = state.db.users
        .find_first(|q| q.where_id(account.id.clone()))
        .await?
        .ok_or(AppError::Unauthorized)?;

    if body.content.is_empty() || body.content.len() > 2000 {
        return Err(AppError::BadRequest("messages.errors.validationFailed".to_string()));
    }

    if channel.rate_limit_per_user > 0 {
        let owner_id = state.guild_owner(&member.guild_id).await?;
        let is_owner = owner_id.as_deref() == Some(account.id.as_str());

        if !is_owner {
            let mut redis = state.redis.clone();
            let key = format!("channel:{}:msg", channel_id);
            let result = rate_limit(
                &mut redis,
                &account.id,
                1,
                (channel.rate_limit_per_user as i64) * 1000,
                &key,
            ).await?;
            if result.limited {
                return Err(AppError::TooManyRequests {
                    message: "messages.errors.rateLimited".to_string(),
                    retry_after: result.retry_after,
                });
            }
        }
    }

    let message_id = generate_snowflake();
    let nonce = body.nonce.clone().unwrap_or_else(|| "0".to_string());
    
    let new_msg = state.db.messages.create(|c| c
        .set_id(message_id.clone())
        .set_author_id(account.id.clone())
        .set_channel_id(channel_id.clone())
        .set_guild_id(member.guild_id.clone())
        .set_content(Some(body.content.clone()))
        .set_message_type(MessageType::DEFAULT)
        .set_nonce(nonce.clone())
    ).await.map_err(|e| AppError::InternalServerError(e.to_string()))?;

    let mut response: Message = new_msg.into();
    response.author = Some(User::from(user));
    
    let broadcast_message = json!({
        "op": 0,
        "t": "MESSAGE_CREATE",
        "d": response
    });
    
    state.broadcast_to_room(&member.guild_id, broadcast_message.to_string());
    
    Ok(Json(response))
}

async fn update_message(
    State(state): State<SharedState>,
    jar: CookieJar,
    Path((channel_id, message_id)): Path<(String, String)>,
    body: axum::body::Bytes,
) -> Result<impl IntoResponse, AppError> {
    let body: MessageUpdateRequest = serde_json::from_slice(&body)
        .map_err(|_| AppError::BadRequest("Invalid JSON".to_string()))?;
    let account = get_current_user(&state, &jar).await?;
    let (_channel, _member) = get_channel_and_member(&state, &channel_id, &account.id).await?;
    
    let message = state.db.messages
        .find_first(|q| q.where_id(message_id.clone()))
        .await?
        .ok_or(AppError::NotFound("messages.errors.messageNotFound".to_string()))?;
    
    if message.author_id != account.id {
        return Err(AppError::Forbidden("messages.errors.notYourMessage".to_string()));
    }
    
    if let Some(content) = &body.content {
        if message.content.as_ref() == Some(content) {
            let response: Message = message.into();
            return Ok(Json(response));
        }
        
        state.db.messages
            .update(|u| u
                .set_content(Some(content.clone()))
                .set_updated_at(Some(chrono::Utc::now()))
                .where_id(message_id.clone())
            )
            .await?;
    }
    
    let updated = state.db.messages
        .find_first(|q| q.where_id(message_id))
        .await?
        .ok_or(AppError::InternalServerError("Failed to update message".to_string()))?;

    let author = state.db.users
        .find_first(|q| q.where_id(updated.author_id.clone()))
        .await?;

    let mut response: Message = updated.clone().into();
    response.author = author.map(User::from);

    let broadcast_message = json!({
        "op": 0,
        "t": "MESSAGE_UPDATE",
        "d": response
    });

    state.broadcast_to_room(&updated.guild_id, broadcast_message.to_string());

    Ok(Json(response))
}

async fn delete_message(
    State(state): State<SharedState>,
    jar: CookieJar,
    Path((channel_id, message_id)): Path<(String, String)>,
) -> Result<impl IntoResponse, AppError> {
    let account = get_current_user(&state, &jar).await?;
    let (_channel, _member) = get_channel_and_member(&state, &channel_id, &account.id).await?;
    
    let message = state.db.messages
        .find_first(|q| q.where_id(message_id.clone()))
        .await?
        .ok_or(AppError::NotFound("messages.errors.messageNotFound".to_string()))?;
    
    if message.author_id != account.id {
        return Err(AppError::Forbidden("messages.errors.forbidden".to_string()));
    }
    
    state.db.messages
        .delete(|d| d.where_id(message_id.clone()))
        .await?;
    
    let broadcast_message = json!({
        "op": 0,
        "t": "MESSAGE_DELETE",
        "d": {
            "id": message.id,
            "channelId": message.channel_id,
            "guildId": message.guild_id
        }
    });
    
    state.broadcast_to_room(&message.guild_id, broadcast_message.to_string());
    
    Ok(StatusCode::NO_CONTENT)
}

async fn typing_start(
    State(state): State<SharedState>,
    jar: CookieJar,
    Path(channel_id): Path<String>,
) -> Result<impl IntoResponse, AppError> {
    let account = get_current_user(&state, &jar).await?;
    let (_channel, member) = get_channel_and_member(&state, &channel_id, &account.id).await?;
    
    let user = state.db.users
        .find_first(|q| q.where_id(account.id.clone()))
        .await?
        .ok_or(AppError::Unauthorized)?;
    
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64;
    
    let broadcast_message = json!({
        "op": 0,
        "t": "TYPING_START",
        "d": {
            "channelId": channel_id,
            "userId": account.id,
            "user": {
                "id": user.id,
                "username": user.username,
                "tag": user.tag,
                "avatar": user.avatar,
                "member": {
                    "nickname": member.nickname
                }
            },
            "timestamp": now,
            "expiresAt": now + 10000
        }
    });
    
    state.broadcast_to_room(&member.guild_id, broadcast_message.to_string());
    
    Ok(Json(json!({"success": true})))
}

async fn create_invite(
    State(state): State<SharedState>,
    jar: CookieJar,
    Path(channel_id): Path<String>,
    body: axum::body::Bytes,
) -> Result<impl IntoResponse, AppError> {
    let body: InviteCreateRequest = serde_json::from_slice(&body)
        .map_err(|_| AppError::BadRequest("Invalid JSON".to_string()))?;
    let account = get_current_user(&state, &jar).await?;
    let (_channel, member) = get_channel_and_member(&state, &channel_id, &account.id).await?;
    
    let code = generate_code();
    let invite_id = generate_snowflake();
    let max_uses = body.max_uses.unwrap_or(0);
    
    state.db.guild_invites
        .create(|c| c
            .set_id(invite_id.clone())
            .set_guild_id(member.guild_id.clone())
            .set_code(code.clone())
            .set_uses(0)
            .set_max_uses(max_uses)
            .set_creator_id(account.id.clone())
            .set_channel_id(channel_id.clone())
            .set_vanity(false)
        )
        .await?;
    
    let invite = state.db.guild_invites
        .find_first(|q| q.where_id(invite_id))
        .await?
        .ok_or(AppError::InternalServerError("Failed to create invite".to_string()))?;
    
    Ok(Json(Invite {
        id: invite.id,
        guild_id: invite.guild_id,
        code: invite.code,
        uses: invite.uses,
        max_uses: invite.max_uses,
        creator_id: Some(invite.creator_id),
        channel_id: invite.channel_id,
        vanity: invite.vanity,
    }))
}
